pkg/api/plugins.go

```diff
@@ -235,6 +237,48 @@ func (hs *HTTPServer) CollectPluginMetrics(c *models.ReqContext) response.Respon
 	return response.CreateNormalResponse(headers, resp.PrometheusMetrics, http.StatusOK)
 }
 
+// GetPluginAssets returns public plugin assets (images, JS, etc.)
+//
+// /public/plugins/:pluginId/*
+func (hs *HTTPServer) GetPluginAssets(c *models.ReqContext) {
+	pluginID := c.Params("pluginId")
+	plugin := hs.PluginManager.GetPlugin(pluginID)
+	if plugin == nil {
+		c.Handle(hs.Cfg, 404, "Plugin not found", nil)
+		return
+	}
+
+	requestedFile := filepath.Clean(c.Params("*"))
+	pluginFilePath := filepath.Join(plugin.PluginDir, requestedFile)
+
+	// It's safe to ignore gosec warning G304 since we already clean the requested file path and subsequently
+	// use this with a prefix of the plugin's directory, which is set during plugin loading
+	// nolint:gosec
+	f, err := os.Open(pluginFilePath)
+	if err != nil {
+		c.Handle(hs.Cfg, 500, "Could not open plugin file", err)
+		return
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			hs.log.Error("Failed to close file", "err", err)
+		}
+	}()
+
+	fi, err := f.Stat()
+	if err != nil {
+		c.Handle(hs.Cfg, 500, "Plugin file exists but could not open", err)
+		return
+	}
+
+	if shouldExclude(fi) {
+		c.Handle(hs.Cfg, 404, "Plugin file not found", nil)
+		return
+	}
+
+	http.ServeContent(c.Resp, c.Req.Request, pluginFilePath, fi.ModTime(), f)
```

Comment:
One thing missing from the earlier use of `mapStatic` is the `Cache-Control` header which I think is important to provide a good experience for end users in browsers not having to re-download plugin assets over and over again. Ref:
https://github.com/grafana/grafana/blob/d807fbc9e9e8dead6c8d20494639703fc0d4090b/pkg/api/http_server.go#L447-L472

Reply:
Yeah I had this originally and then removed as I'm not sure how caching will impact plugin updates for example - will files be potentially wrongfully cached in that case?

Reply:
I created a branch with some modifications since it was too big trying to explain it in a comment. A possible way to handle static resources using a middleware that will set context key and provide cache headers. With the context key set we can use that to exclude certain middleware from executing. WDYT?  https://github.com/marefr/grafana/commit/b767b41283ae97ba3b513da62a2e9fe15f654e6f

Reply:
Checking how Grafana loads plugin assets it adds a cache buster query string parameter like http://localhost:3000/public/plugins/alexanderzobnin-zabbix-app/module.js?_cache=1618850957042. It's defined here https://github.com/grafana/grafana/blob/d807fbc9e9e8dead6c8d20494639703fc0d4090b/public/app/features/plugins/plugin_loader.ts#L50

So excluding plugin assets from adding no cache headers is probably not needed :smile: